### PR TITLE
Surface + document optimizer/trainer defaults (opt-in) + fractional warmup

### PR DESCRIPTION
**Pull request by farhanferoz**
## Summary

Surfaces four training-config knobs that were previously absent or latent, and ships the tuned values as **opt-in selectable configs** — the live defaults are left unchanged, so existing runs reproduce from config alone and old W&B baselines stay comparable.

Opt in with:

```
optimizer=updated_defaults/adamw   # weight_decay 0.01, warmup 0.05
trainer=updated_defaults           # gradient_clip_val 1.0 (inherits trainer/default)
```

| Setting | Live default (unchanged) | `updated_defaults` (opt-in) |
|---|---|---|
| `adamw.yaml` `weight_decay` | `0.0` | `0.01` |
| `adamw.yaml` `warmup` | `0` | `0.05` (fraction of cosine horizon) |
| `trainer.gradient_clip_val` | `null` (no clipping) | `1.0` |
| `trainer.precision` | unset → `32-true` | unchanged — bf16 dropped (see below) |

Plus fractional-warmup support in `OptimizerMixin._resolve_warmup`: a float in `(0, 1)` is a fraction of the cosine horizon; integers and floats `>= 1` stay absolute counts; `NaN`/`inf` rejected.

## Approach (per review)

Following @sgreenbury's review: rather than changing defaults in place, the tuned values live in selectable `updated_defaults` configs, so nothing changes for existing runs. #394 tracks promoting them to the in-place defaults once validated.

## bf16-mixed dropped

Not recommended — earlier fp32-vs-bf16 experiments showed degradation ([report](https://wandb.ai/turing-core/autocast/reports/Comparison-of-fp32-and-bf16-mixed--VmlldzoxNjUzNTc0Nw)). Precision stays at Lightning's `32-true`. bf16 is still usable as a manual override, and a smoke test keeps it numerically exercised (no NaN/Inf), but it is not part of the recommended set.

## Migration: `optimizer.grad_clip` removed

The orphaned `grad_clip` field is dropped from all optimizer configs, and `_create_optimizer` now **raises** if `optimizer.grad_clip` is set, pointing callers to `trainer.gradient_clip_val`. This is deliberate: any config that set `optimizer.grad_clip: <value>` was silently *not* clipping (no code path read it). Such configs must move the value to `trainer.gradient_clip_val` — or drop it, since it was a no-op. No config in this repo sets it; only downstream/experiment configs need the one-line move.

## References

The values are framework/community defaults aligned to the LOLA reference, not paper-derived constants; the citations justify the techniques and explain why they help (validation before promotion is tracked in #394).

- **Weight decay (0.01)** — decoupled WD: Loshchilov & Hutter, ICLR 2019 (arXiv:1711.05101); modern role (optimisation dynamics, not classical regularisation): D'Angelo et al., NeurIPS 2024 (arXiv:2310.04415). 0.01 is the PyTorch AdamW default.
- **Warmup (5%)** — mechanism: Kalra & Barkeshli, NeurIPS 2024 (arXiv:2406.09405); Adam-specific variance argument: Liu et al. (RAdam), ICLR 2020 (arXiv:1908.03265); linear-warmup origin: Goyal et al., 2017 (arXiv:1706.02677).
- **Cosine schedule** — Loshchilov & Hutter, SGDR, ICLR 2017 (arXiv:1608.03983); match the cycle length to the run: Hoffmann et al. (Chinchilla), 2022 (arXiv:2203.15556).
- **Gradient clipping (1.0)** — why it accelerates: Zhang et al., ICLR 2020 (arXiv:1905.11881); origin: Pascanu et al., ICML 2013 (arXiv:1211.5063). 1.0 is the GPT-style global-norm default and matches LOLA.

## Notes

- **WD `0.01` is a starting point.** `0.05` is the ViT-*classification* recipe; this is field regression where it tends to over-regularise. When enabled it is applied uniformly — no no-decay param groups for biases/norm scales yet (tracked in #394).
- **`adamw_half.yaml` stays LOLA-faithful** (`weight_decay: 0.0`, `warmup: 0`).
- Relates to #322 (optimizer/trainer param alignment).

## Testing

- 16 unit tests for `_resolve_warmup`; 3 for the `grad_clip` migration guard; a bf16-mixed + `gradient_clip_val=1.0` autoencoder fit smoke; a config pin asserting the reproducibility-preserving live defaults; a compose test asserting the `updated_defaults` configs select cleanly, carry the tuned values, and preserve inherited trainer callbacks.
- `uv run pytest tests/models/test_optimizer_mixin.py tests/scripts/test_training.py` → 56 passed. Ruff + pyright clean.

### CODE REVIEW

**Review by sgreenbury — COMMENTED** [review]
Thanks for adding this @farhanferoz; updating the defaults to better/widely used ones sounds good!

A few comments as we discussed:
- Re. the behaviour change - perhaps we can copy the new configs and move to a subpath (e.g. `configs/optimizer/updated_defaults`?) and we open a new issue to then replace the current ones
- It might be worth not switching to `bf16-mixed` for the moment; we saw some perf degradation when experimenting with this before ([wandb](https://wandb.ai/turing-core/autocast/reports/Comparison-of-fp32-and-bf16-mixed--VmlldzoxNjUzNTc0Nw)).
- Adding some references in the literature for the new defaults could be useful

**Comment by farhanferoz** [comment]
Thanks @sgreenbury! Addressed all three:

1. **Subpath + follow-up issue** — done. Tuned values now live as opt-in configs (`optimizer=updated_defaults/adamw`, `trainer=updated_defaults`); the live defaults are untouched. Opened #394 to track promoting them once validated.
2. **bf16-mixed** — dropped from the recommended set given the degradation in your report; precision stays `32-true`. Kept a smoke test so the option stays numerically exercised, but it's no longer recommended.
3. **References** — added foundational + recent citations to the configs, the PR description, and #394: decoupled WD (Loshchilov & Hutter 2019; D'Angelo et al. NeurIPS 2024), warmup (Kalra & Barkeshli NeurIPS 2024; RAdam), cosine (SGDR; Chinchilla on matching the horizon to the run), clipping (Zhang et al. ICLR 2020). The values themselves are framework/LOLA defaults — the refs justify the techniques; validating the numbers is tracked on #394.

The `grad_clip` rewire stays (orphaned `optimizer.grad_clip` → `trainer.gradient_clip_val` + migration guard) since that was a genuine no-op bug; flagged broader alignment on #322.

**Review by sgreenbury — APPROVED** [review]
Looks great, thanks @farhanferoz!
